### ISSUE TITLE
Give KHR_vulkan_glsl push constant blocks an `offset` qualifier.

### ISSUE THREAD

**Issue by NicolBolas** [KhronosGroup/Vulkan-Docs#435]
Pursuant to issue #255, I'm making a request that KHR_vulkan_glsl include the `offset` qualifier in certain situations, mainly so that the extension can more effectively be applied to ESSL versions. No version of ESSL permits the `offset` on uniform or storage blocks, and some of Vulkan's features rely strongly on such things.

The principle problem is dealing with push constant ranges. Some code that tries to share source between ESSL and Vulkan might do something such as the following:

````
layout(
#ifdef VULKAN
  push_constant, 
#else
  binding = 0,
#endif
std140) uniform PCBlock
{
  //Members
} pc_instance;
````

The problem is that the offset for the members of the push-constant block will start at 0, per the normal layout rules. But different shader stages may need distinct push-constant ranges. And that means being able to specify the initial offset of a range.

In desktop GLSL, one could use GLSL 4.40 or ARB_enhanced_layouts, which permits users to apply the `offset` layout qualifier to block members. But no version of ESSL has this functional, nor does any extension.

We could resolve this by making that portion of ARB_enhanced_layouts part of KHR_vulkan_glsl. But this puts the `offset` qualifier on the first block member definition. Instead, we could allow `offset` to be part of the *block*'s layout qualifier, but only if it also uses the `push_constant` layout:

````
layout(
#ifdef VULKAN
  push_constant, offset = 64,
#else
  binding = 0,
#endif
std140) uniform PCBlock
{
  //Members
} pc_instance;
````

This grammar is not part of any GLSL version, so we can give it whatever meaning we like. For example, we can declare that any member `offset` layouts are relative to the block's overall base `offset`.

**Comment by johnkslang** [comment]
If I'm understanding, is this two feature requests?

1. allow `offset` on a block (today only allowed on members)
2. allow `offset` for all versions (today, only allowed on some versions)

I.e., if we did feature 2, would we still need feature 1?

```
layout(
#ifdef VULKAN
  push_constant,
  binding = 0,
#endif
std140) uniform PCBlock
{
    #ifdef VULKAN
      layout(offset = 64)
    #endif
    float member;
} pc_instance;
```

This is just a clarifying question, thanks.  I think feature 2 is a pretty simple language issue, while feature 1 is potentially more complex (and possibly) a Vulkan issue.

**Comment by NicolBolas** [comment]
I hadn't realized that `offset` only could be applied to block members, not to the blocks themselves. Since that's already part of GLSL, feature 2 should be adequate.

**Comment by johnkslang** [comment]
We are discussing this now in Khronos: allowing `offset` for all versions when compiling for Vulkan.

**Comment by johnkslang** [comment]
Khronos has approved adding `offset` and `align` block-member functionality, independent of `#version` when compiling for Vulkan. Will update glslang accordingly.

**Comment by oddhack** [comment]
This should be fixed in the 1.0.41 spec update.
